README.md
# Semantic UI for Sass

`semantic-ui-sass` is an Sass-powered version of
[Semantic UI](https://github.com/Semantic-Org/Semantic-UI) and ready to drop
into Rails, Compass, or Sprockets.

[![Build Status](https://travis-ci.com/doabit/semantic-ui-sass.svg?branch=master)](https://travis-ci.com/doabit/semantic-ui-sass)

## NOTE

The gem only has default theme.

## Installation and Usage

```ruby
gem 'semantic-ui-sass'
```

`bundle install` and restart your server to make the files available through the
pipeline.

# semantic-ui-sass with Rails or Sprockets

## CSS

Import Semantic in an SCSS file (for example, `application.css.scss`) to get all
of Semantic's styles

```css
@import 'semantic-ui';
```

You can also include modules

```css
@import 'semantic-ui/collections/menu';
```

## Custom font

```css
$font-url: 'http://fonts.useso.com/css?family=Lato:400,700,400italic,700italic&subset=latin';
@import 'semantic-ui';
```

## Skip font loading

```css
$import-google-fonts: false;
@import 'semantic-ui';
```

## Custom font family

```css
$font-family: 'custom-font-family';
@import 'semantic-ui';
```

### Skip use of custom scrollbars

```css
$use-custom-scrollbars: false;
@import 'semantic-ui';
```

## All variables, you can custom any of that

Fonts:

```css
$font-name: 'Lato' !default;
$font-family: $font-name, 'Helvetica Neue', Arial, Helvetica, sans-serif !default;
$font-url: 'https://fonts.googleapis.com/css?family=Lato:400,700,400italic,700italic&subset=latin,latin-ext' !default;
$import-google-fonts: true !default;
```

Scrollbars:

```css
$use-custom-scrollbars: true !default;
```

Breakpoints:

```css
$mobile-breakpoint: 320px !default;
$tablet-breakpoint: 768px !default;
$computer-breakpoint: 992px !default;
$large-monitor-breakpoint: 1200px !default;
$widescreen-monitor-breakpoint: 1920px !default;
```

## Javascripts

### Ruby on Rails Version 5

We have a helper that includes all Semantic javascripts. Put this in your
Javascript manifest (usually in `application.js`) to

```js
// Loads all Semantic javascripts
//= require semantic-ui
```

You can also load individual modules, provided you also require any
dependencies.

```js
//= require semantic-ui/modal
//= require semantic-ui/dropdown
```

### Ruby on Rails Version 6+ with webpacker

Add packages with yarn:

```console
yarn add jquery popper.js @doabit/semantic-ui-sass
```

In config/webpack/environment.js add the following:

```js
const webpack = require('webpack')

environment.plugins.append(
  'Provide',
  new webpack.ProvidePlugin({
    $: 'jquery',
    jQuery: 'jquery',
    Popper: ['popper.js', 'default']
  })
)
```

In app/javascript/packs/application.js add the following:

```js
import '@doabit/semantic-ui-sass'
```

If you use `webpakcer` to import stylesheet, you should add

```js
import '@doabit/semantic-ui-sass/src/scss/semantic-ui.scss'
```

[Rails6 webpacker demo](https://github.com/doabit/semantic-ui-sass-rails6-demo)

# semantic-ui-sass with Compass

## New project

Install the gem and create a new project using the gem.

```console
gem install semantic-ui-sass
compass create compass-project -r semantic-ui-sass --using semantic-ui
```

This will sort a few things out:

- You'll get a starting `styles.scss`
- You'll get a compiled stylesheet compiled & ready to drop into your
  application
- We'll also copy the Semantic javascripts & images & fonts into their
  respective folders for you

## Existing project

Install the gem, add the require statement to the top of your configuration
file, and install the extension.

```console
gem install semantic-ui-sass
```

```ruby
# In config.rb
require 'semantic-ui-sass'
```

```console
compass install semantic-ui
```

### NOTE

When using compass, you should visit file in local server, eg
`http://localhost:3000/index.html`, rather than
`file:///Users/doabit/demo/index.html`

# Rails Helpers

## Breadcrumbs helper

Add breadcrumbs helper `<%= semantic_breadcrumbs %>` to your layout.

```ruby
class ApplicationController
  semantic_breadcrumb :index, :root_path
end
```

```ruby
class ExamplesController < ApplicationController
  semantic_breadcrumb :index, :examples_path

  def index
  end

  def show
    @example = Example.find params[:id]
    semantic_breadcrumb @example.name, example_path(@example)
    # semantic_breadcrumb :show, example_path(@example)
  end
end
```

## Flash helper

Add flash helper `<%= semantic_flash %>` to your layout

## Icon helper

```ruby
semantic_icon('add')
# => <i class="add icon"></i>
semantic_icon(:add)
# => <i class="add icon"></i>
semantic_icon('add sign')
# => <i class="add sign icon"></i>
semantic_icon('add', 'sign')
# => <i class="add sign icon"></i>
semantic_icon(:add, :sign)
# => <i class="add sign icon"></i>
semantic_icon('add', id: 'id')
# => <i class="add icon" id="id"></i>
```

## TODO

- Add global variables
- Add rails helpers like `render_flash`?

## Contributing

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request


CHANGELOG.md
## 2.4.4.0

- Add npm package

## 2.4.3.0

- Change from `sass-rails` to `sassc-rails` to correct deprecation usage

## 2.4.2.0

- Update Semantic UI to 2.4.2
- Add $use-custom-scrollbar variable to optionally use native scrollbars #138

## 2.4.0.1

- Include new placeholder element #133

## 2.4.0.0

- Update Semantic UI to 2.4.0

## 2.3.3.0

- Update Semantic UI to 2.3.3

## 2.3.1.2

- update icons font

## 2.3.1.1

- add missing brand-icon files #128

## 2.3.1.0

- Update Semantic UI to 2.3.1

## 2.3.0.0

- Update Semantic UI to 2.3.0

## 2.2.14

- Update Semantic UI to 2.2.14

## 2.2.12.1

- Add sprockets support #120

## 2.2.10.1

- Fix flag.png path

## 2.2.10.0

- Update Semantic UI to 2.2.10

## 2.2.9.3

- Fix semantic-ui.js require statements

## 2.2.9.2

- Update semantic-ui.js to 2.2.9

## 2.2.9.1

- Fix compass template error

## 2.2.9.0

- before_action instead before_filter

- Update Semantic UI to 2.2.9

## 2.2.7.1

- Fix invalid image-url for flags.png

## 2.2.7.0

- Replace flags image with optimized one

- Change gem install to secure HTTPS URL

- Update Semantic UI to 2.2.7

## 2.2.6.0

- Add import-google-fonts variable

- Update Semantic UI to 2.2.6

## 2.2.4.0

- Update Semantic UI to 2.2.4

## 2.2.3.0

- Add font-name variable to change easily custom fonts
- Fix undefined method `start_with?` for sprockets 4
- Add font-family variable
- Update Semantic UI to 2.2.3

## 2.2.2.2

- Update fonts to version 2.2.2 #81

## 2.2.2.1

- Add font-url variable

## 2.2.2.0

- Update Semantic UI to 2.2.2

## 2.2.1.1

- Fix #79

## 2.2.1.0

- Update Semantic UI to 2.2.1
- Fixed font url error #78

## 2.2.0.0

- Update Semantic UI to 2.2.0
- Fixed rails 5 error
- Allow any HTML options in semantic_icon
- Add latin-ext subset to Google Font import URL.

## 2.1.8.0

- Update Semantic UI to 2.1.8

## 2.1.6.0

- Update Semantic UI to 2.1.6

## 2.1.4.0

- Rename breadcrumb methods to follow semantic\_ naming

- Update Semantic UI to 2.1.4

## 2.0.7.0

- Update Semantic UI to 2.0.7

## 2.0.4.0

- Update Semantic UI to 2.0.4

## 1.12.3.0

- Update Semantic UI to 1.12.3

## 1.11.4.1

- Removes reference to a absent icon file

## 1.11.4.0

- Update Semantic UI to 1.11.4

## 1.8.1.0

- Update Semantic UI to 1.8.1

## 1.8.0.0

- Update Semantic UI to 1.8.0

## 1.7.3.0

- Update Semantic UI to 1.7.3

## 1.7.0.0

- Update Semantic UI to 1.7.0

## 1.6.2.0

- Update Semantic UI to 1.6.2

## 1.5.0.0

- Update Semantic UI to 1.5.0

## 1.4.1.0

- Update Semantic UI to 1.4.1

## 1.2.0.0

- Update Semantic UI to 1.2.0

## 0.19.3.1

- Fix fonts urls for asset pipeline

## 0.19.3.0

- Fix font-url

- Update Semantic UI to 0.19.3

## 0.16.1.0

- Update Semantic UI to 0.16.1

## 0.15.5.0

- Update Semantic UI to 0.15.5

## 0.15.4.2

- Fix netsted import error

## 0.15.4.1

- Fix depend_on_asset

## 0.15.4.0

- Update Semantic UI to 0.15.4

## 0.15.2.0

- Update Semantic UI to 0.15.2

## 0.15.1.0

- Update Semantic UI to 0.15.1

## 0.14.0.0

- Update Semantic UI to 0.14.0

## 0.13.1.0

- Update Semantic UI to 0.13.1

## 0.13.0.0

- Update Semantic UI to 0.13.0

## 0.12.5.0

- Update Semantic UI to 0.12.5

## 0.12.4.0

- Update Semantic UI to 0.12.4

## 0.12.3.0

- Update Semantic UI to 0.12.3

## 0.12.2.0

- Update Semantic UI to 0.12.2

## 0.12.1.0

- Update Semantic UI to 0.12.1

## 0.12.0.0

- Update Semantic UI to 0.12.0

## 0.11.0.0

- Update Semantic UI to 0.11.0

## 0.10.3.0

- Update Semantic UI to 0.10.3

## 0.10.2.0

- Update Semantic UI to 0.10.2

## 0.10.1.0

- Update Semantic UI to 0.10.1

## 0.10.0.0

- Update Semantic UI to 0.10.0

## 0.9.6.0

- Update Semantic UI to 0.9.6

## 0.9.5.0

- Update Semantic UI to 0.9.5

## 0.9.4.0

- Update Semantic UI to 0.9.4

## 0.9.3.0

- Update Semantic UI to 0.9.3

## 0.9.2.0

- Update Semantic UI to 0.9.2

## 0.9.1.0

- Update Semantic UI to 0.9.1

- Add breadcrumbs helper

## 0.9.0.0

- Update Semantic UI to 0.9.0

- Add browsers setting for autoprefixer

## 0.8.6.0

- Update Semantic UI to 0.8.6

## 0.8.5.0

- Update Semantic UI to 0.8.5

## 0.8.4.0

- Update Semantic UI to 0.8.4

## 0.8.3.0

- Update Semantic UI to 0.8.3

## 0.8.0.0

## 0.8.2.0

- Update Semantic UI to 0.8.2

- Add autoprefixer

- Remove sqlite3

## 0.8.1.0

- Update Semantic UI to 0.8.1

## 0.8.0.0

- Add flash helper

- Add icon helper

- Update Semantic UI to 0.8.0

## 0.0.3

- Fix typo in semantic-ui.js

- Update Semantic UI

## 0.0.2

### Features

- Add support for compass

```console
  gem install semantic-ui-sass
  compass create compass-project -r semantic-ui-sass --using semantic-ui
```

## 0.0.1

- Initial version


tasks/converter.rb
# Based on convert script from vwall/compass-twitter-bootstrap gem.
# https://github.com/vwall/compass-twitter-bootstrap/blob/#{@branch}/build/convert.rb

require 'open-uri'
require 'json'
require 'fileutils'
require "pry"
require "dotenv"

require 'pathname'

Dotenv.load

class Converter

  GIT_DATA = 'https://api.github.com/repos'
  GIT_RAW  = 'https://raw.githubusercontent.com'
  TOKEN    = ENV['TOKEN']


  def initialize(branch)
    @repo               = 'Semantic-Org/Semantic-UI'
    @repo_url           = "https://github.com/#@repo"
    @branch             = branch || 'master'
    @git_data_trees     = "#{GIT_DATA}/#{@repo}/git/trees"
    @git_raw_src        = "#{GIT_RAW}/#{@repo}/#{@branch}/dist"
    @branch_sha         = get_tree_sha
  end

  def process
    # process_stylesheets_assets

    process_images_and_fonts_assets
    # store_version

    checkout_repository
    choose_version(@branch)
    process_stylesheets_assets
    process_javascript_assets
    store_version
  end

  def paths
    @gem_paths ||= Paths.new
  end

  def checkout_repository
    if Dir.exist?(paths.tmp_semantic_ui)
      system %Q{cd '#{paths.tmp_semantic_ui}' && git fetch --quiet}
    else
      system %Q{git clone --quiet git@github.com:Semantic-Org/Semantic-UI.git '#{paths.tmp_semantic_ui}'}
    end
  end

  def choose_version(version)
    system %Q{cd '#{paths.tmp_semantic_ui}' && git checkout --quiet #{version}}
  end

  def process_images_and_fonts_assets
    fonts = File.join(paths.tmp_semantic_ui_dist, 'themes/default/assets/fonts', '*')
    Dir[fonts].each do |src|
      FileUtils.cp(src, paths.fonts)
    end
    images = File.join(paths.tmp_semantic_ui_dist, 'themes/default/assets/images', '*')
    Dir[images].each do |src|
      FileUtils.cp(src, paths.images)
    end
  end

  def process_stylesheets_assets
          # content = ""
    Dir[File.join(paths.tmp_semantic_ui_definitions, '*')].each do |path|
      # all = ""

      Dir[File.join(path, "*.less")].each do |file|

        filename = File.basename(file).gsub(".less", '.css')

        Dir[File.join(paths.tmp_semantic_ui_components, "*.css")].each do |src|
          name = File.basename(src)

         if name == filename
           file = open(src).read
           file = convert(file)
           save_file(name, file, File.basename(path))

           # all << "@import '#{name.gsub(".css", "")}';\n"
         end #filename
        end
      end
      # save_file("all", all, File.basename(path)) if all != ''
      # content << "@import 'semantic-ui/#{File.basename(path)}/all';\n";

    end
     # File.open("app/assets/stylesheets/semantic-ui.scss", "w+") { |file| file.write(content) }
  end


  def process_javascript_assets
    # js = ""
    Dir[File.join(paths.tmp_semantic_ui_definitions, '**/*.js')].each do |src|
       name = File.basename(src).gsub(".js", '')
       # js << "//= require #{name}\n"
       FileUtils.cp(src, paths.javascripts)
     end
     # File.open("app/assets/javascripts/semantic-ui.js", "w+") { |file| file.write(js) }
  end


private

  # Get the sha of less branch
  def get_tree_sha
    sha = nil
    trees = get_json("#{@git_data_trees}/#{@branch}")
    trees['tree'].find{|t| t['path'] == 'dist'}['sha']
  end


  def convert(file)
    file = replace_fonts_url(file)
    file = replace_import_font_url(file)
    file = replace_font_family(file)
    file = replace_image_urls(file)
    file = replace_image_paths(file)
    file = add_breakpoint_variables(file)

    file
  end


  def save_file(name, content, path, type='stylesheets')

    name = name.gsub(/\.css/, '')
    file = "#{paths.stylesheets}/#{path}/_#{name}.scss"
    dir = File.dirname(file)
    FileUtils.mkdir_p(dir) unless File.directory?(file)
    File.open(file, 'w+') { |f| f.write(content) }
    # puts "Saved #{name} at #{path}\n"
  end



  def get_json(url)
    url += "?access_token=#{TOKEN}" unless TOKEN.nil?
    data = open_git_file(url)
    data = JSON.parse data
  end

  def open_git_file(file)
    open(file).read
  end

  def store_version
    path = 'lib/semantic/ui/sass/version.rb'
    content = File.read(path).sub(/SEMANTIC_UI_SHA\s*=\s*['"][\w]+['"]/, "SEMANTIC_UI_SHA = '#@branch_sha'")
    File.open(path, 'w') { |f| f.write(content) }
  end

  def replace_fonts_url(less)
    less.gsub(/url\(\"\.\/\.\.\/themes\/default\/assets\/fonts\/?(.*?)\"\)/) {|s| "font-url(\"semantic-ui/#{$1}\")" }
  end

  def replace_font_family(less)
    less.gsub("font-family: 'Lato', 'Helvetica Neue', Arial, Helvetica, sans-serif", 'font-family: $font-family')
  end

  def replace_import_font_url(less)
    less.gsub("@import url('https://fonts.googleapis.com/css?family=Lato:400,700,400italic,700italic&subset=latin');", '@if $import-google-fonts {
  @import url($font-url);
}')
  end

  def replace_image_urls(less)
    less.gsub(/url\("?(.*?).png"?\)/) {|s| "image-url(\"#{$1}.png\")" }
  end

  def replace_image_paths(less)
    less = less.gsub('./../themes/default/assets/images/flags.png', 'semantic-ui/flags.png')
    less.gsub('../themes/default/assets/images/', 'semantic-ui/')
  end

  def add_breakpoint_variables(css)
    css = css.gsub('(min-width: 320px)', '(min-width: $mobile-breakpoint)')
    css = css.gsub('max-width: 767px', 'max-width: $largest-mobile-screen')

    css = css.gsub('min-width: 768px', 'min-width: $tablet-breakpoint')
    css = css.gsub('max-width: 991px', 'max-width: $largest-tablet-screen')

    css = css.gsub('min-width: 992px', 'min-width: $computer-breakpoint')
    css = css.gsub('max-width: 1199px', 'max-width: $largest-small-monitor')

    css = css.gsub('min-width: 1200px', 'min-width: $large-monitor-breakpoint')
    css = css.gsub('max-width: 1919px', 'max-width: $largest-large-monitor')

    css = css.gsub('min-width: 1920px', 'min-width: $widescreen-monitor-breakpoint')
  end
end

class Paths
   attr_reader :root
   attr_reader :tmp
   attr_reader :tmp_semantic_ui
   attr_reader :tmp_semantic_ui_src
   attr_reader :tmp_semantic_ui_definitions
   attr_reader :tmp_semantic_ui_dist
   attr_reader :tmp_semantic_ui_components

   attr_reader :fonts
   attr_reader :images
   attr_reader :javascripts
   attr_reader :stylesheets


   def initialize
     @root = File.expand_path('..', __dir__)

     @tmp = File.join(@root, 'tmp')
     @tmp_semantic_ui = File.join(@tmp, 'semantic-ui')
     @tmp_semantic_ui_src = File.join(@tmp_semantic_ui, 'src')
     @tmp_semantic_ui_definitions = File.join(@tmp_semantic_ui_src, 'definitions')

     @tmp_semantic_ui_dist = File.join(@tmp_semantic_ui, 'dist')
     @tmp_semantic_ui_components = File.join(@tmp_semantic_ui_dist, 'components')

     @app = File.join(@root, 'app')
     @fonts = File.join(@app, 'assets', 'fonts', 'semantic-ui')
     @images = File.join(@app, 'assets', 'images', 'semantic-ui')
     @javascripts = File.join(@app, 'assets', 'javascripts', 'semantic-ui')
     @stylesheets = File.join(@app, 'assets', 'stylesheets', 'semantic-ui')
   end
 end
